File: archives/cpp_gui/menu/data/file_contents.cpp
#include "file_contents.h"


MenuDataFileContents::MenuDataFileContents(QWidget* parent)
    : IodeSettingsDialog(parent)
{
    setupUi(this);

    QStringList listIodeType;
    for(const std::string& iode_type : v_iode_types) listIodeType << QString::fromStdString(iode_type);

    wSearch = new WrapperQLineEdit(label_search->text(), *lineEdit_search, REQUIRED_FIELD);
    wComboIodeType = new WrapperComboBox(label_iode_types->text(), *comboBox_iode_types, REQUIRED_FIELD, listIodeType);
    wInputFile = new WrapperFileChooser(label_input_file->text(), *fileChooser_input_file, OPTIONAL_FIELD, FILE_COMMENTS, EXISTING_FILE);
    wSaveList = new WrapperQLineEdit(label_save_list->text(), *lineEdit_save_list, REQUIRED_FIELD);
    wResult = new WrapperQTextEdit(label_result->text(), *textEdit_result, OPTIONAL_FIELD);

    mapFields["Search"]    = wSearch;
    mapFields["IodeType"] = wComboIodeType;
    mapFields["InputFile"] = wInputFile;
    mapFields["SaveList"]  = wSaveList;

    connect(&wComboIodeType->getQField(), &QComboBox::currentIndexChanged, this, &MenuDataFileContents::updateFileChooser);

    className = "MENU_DATA_FILE_CONTENTS";
    loadSettings();
}

MenuDataFileContents::~MenuDataFileContents()
{
    delete wSearch;
    delete wComboIodeType;
    delete wInputFile;
    delete wSaveList;
    delete wResult;
}

// TODO ALD: implement a file_contents() method in KDBAbstract + tests
// Tests: search = A*, iode type = COMMENTS, file = empty
//        search = A*, iode type = COMMENTS, file = fun2.ac
void MenuDataFileContents::search()
{
    try
    {
        std::string pattern = wSearch->extractAndVerify().toStdString();
        int iode_type = wComboIodeType->extractAndVerify();
        std::string input_file = wInputFile->extractAndVerify().toStdString();
        std::string save_list = wSaveList->extractAndVerify().toStdString();

        std::string buf = save_list + " " + pattern + " " + input_file;
        int res = B_DataList(buf.data(), iode_type);
        if(res < 0)
            B_display_last_error();

        List list = Lists.get(save_list);
        wResult->setQValue(QString::fromStdString(list));
        int nb_elements = B_DataListCount(save_list.data());
        label_nb_elements->setText(QString::number(nb_elements) + " Elements");
    }
    catch(const std::exception& e)
    {
        QMessageBox::warning(nullptr, "WARNING", QString(e.what()));
    }
}

void MenuDataFileContents::help()
{
	QDesktopServices::openUrl(url_manual);
}

File: gui/iode_gui/menu/data/file_contents.py

from qtpy.QtCore import Slot
from qtpy.QtWidgets import QWidget, QMessageBox

from iode_gui.settings import MixinSettingsDialog
from iode_gui.abstract_main_window import AbstractMainWindow
from .ui_file_contents import Ui_MenuDataFileContents

from iode import (IodeType, IodeFileType, comments, equations, identities, 
                  lists, scalars, tables, variables)


class MenuDataFileContents(MixinSettingsDialog):
    def __init__(self, parent: AbstractMainWindow):
        super().__init__(parent)
        self.ui = Ui_MenuDataFileContents()
        self.ui.setupUi(self)
        self.prepare_settings(self.ui)

        self.v_iode_types = list(IodeType)
        v_iode_type_names = [iode_type.name for iode_type in self.v_iode_types]
        self.ui.comboBox_iode_types.addItems(v_iode_type_names)
        self.ui.comboBox_iode_types.setCurrentIndex(0)
        self.update_file_chooser(0)

        self.ui.lineEdit_search.setPlaceholderText("search pattern: A*;*_")
        self.ui.lineEdit_search.setText("_RES")

        self.ui.comboBox_iode_types.currentIndexChanged.connect(self.update_file_chooser)

        self.load_settings()

    @Slot()
    def search(self):
        try:
            pattern: str = self.ui.lineEdit_search.text()
            i_iode_type = self.ui.comboBox_iode_types.currentIndex()
            iode_type: IodeType = self.v_iode_types[i_iode_type]
            input_file: str = self.ui.fileChooser_input_file.filepath
            save_list_name: str = self.ui.lineEdit_save_list.text()

            if pattern == "":
                raise Exception("No search pattern provided.")
            
            if input_file == "":
                input_file = None
            
            if iode_type == IodeType.COMMENTS:
                result_list = comments.get_names(pattern, input_file)
            elif iode_type == IodeType.EQUATIONS:
                result_list = equations.get_names(pattern, input_file)
            elif iode_type == IodeType.IDENTITIES:
                result_list = identities.get_names(pattern, input_file)
            elif iode_type == IodeType.LISTS:
                result_list = lists.get_names(pattern, input_file)
            elif iode_type == IodeType.SCALARS:
                result_list = scalars.get_names(pattern, input_file)
            elif iode_type == IodeType.TABLES:
                result_list = tables.get_names(pattern, input_file)
            elif iode_type == IodeType.VARIABLES:
                result_list = variables.get_names(pattern, input_file)

            if save_list_name:
                lists[save_list_name] = result_list

            self.ui.label_nb_elements.setText(f"{len(result_list)} Elements")
            self.ui.textEdit_result.setText(";".join(result_list))

        except Exception as e:
            self.parent().display_output(f"ERROR -> {str(e)}")
            QMessageBox.warning(self, "WARNING", str(e))

    @Slot(int)
    def update_file_chooser(self, i_iode_type: int):
        iode_type: IodeType = self.v_iode_types[i_iode_type]
        file_type: IodeFileType = IodeFileType(int(iode_type))
        self.ui.fileChooser_input_file.enum_file_type = file_type
